synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -515,18 +515,15 @@ def _get_bare_e2e_cross_signing_keys_bulk_txn(
             clause, params = make_in_list_sql_clause(
                 txn.database_engine, "k.user_id", user_chunk
             )
-            sql = (
-                """
-                SELECT k.user_id, k.keytype, k.keydata, k.stream_id
-                  FROM e2e_cross_signing_keys k
-                  INNER JOIN (SELECT user_id, keytype, MAX(stream_id) AS stream_id
-                                FROM e2e_cross_signing_keys
-                               GROUP BY user_id, keytype) s
-                 USING (user_id, stream_id, keytype)
-                 WHERE
-            """
-                + clause
-            )
+
+            sql = """
+                SELECT DISTINCT ON (user_id, keytype) user_id, keytype, k.keydata, stream_id
```

Comment:
The old query was doing magic with getting the max stream ID, does that matter here or is this equivalent?

Reply:
I've updated it now, but yes postgres does The Right Thing when using `DISTINCT ON` with an `ORDER BY` clause: https://www.postgresql.org/docs/13/sql-select.html#SQL-DISTINCT